test/test_main_invocations.py

```diff
@@ -0,0 +1,136 @@
+# limitations under the License.
+import os
+from pathlib import Path
+from typing import Callable, Type
+from unittest.mock import Mock
+
+import pytest
+
+import ops
+
+Reset = Callable[[], None]
+
+
+def type_test_dummy(_arg: Callable[[Type[ops.CharmBase], bool], None]):
+    """Usage:
+    from somewhere import main
+    type_test_dummy(main)
+    """
+
+
+def type_test_negative(_arg: Callable[[], None]):
+    """Usage:
+    from somewhere import main
+    type_test_negative(main)  # type: ignore
+
+    The `reportUnnecessaryTypeIgnoreComment` setting is expected to kick up a fuss,
+    should the passed argument match the expected argument type.
+    """
+
+
+@pytest.fixture
+def reset(monkeypatch: pytest.MonkeyPatch, tmp_path: Path):
+    monkeypatch.setattr('sys.argv', ('hooks/install',))
+    monkeypatch.setattr('ops._main._emit_charm_event', Mock())
+    monkeypatch.setattr('ops._main._Manager._setup_root_logging', Mock())
+    monkeypatch.setattr('ops.charm._evaluate_status', Mock())
+    monkeypatch.setenv('JUJU_CHARM_DIR', str(tmp_path))
+    monkeypatch.setenv('JUJU_UNIT_NAME', 'test_main/0')
+    monkeypatch.setenv('JUJU_MODEL_NAME', 'mymodel')
+    monkeypatch.setenv('JUJU_DISPATCH_PATH', 'hooks/install')
+    monkeypatch.setenv('JUJU_VERSION', '3.5.0')
+    (tmp_path / 'metadata.yaml').write_text('name: test', encoding='utf-8')
+    (tmp_path / 'dispatch').absolute().touch(mode=0o755)
+
+    yield (reset := lambda: os.environ.pop('OPERATOR_DISPATCH', None))
+
+    reset()
+
+
+class IdleCharm(ops.CharmBase):
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+
+
+def test_top_level_import(reset: Reset):
+    import ops
+
+    type_test_dummy(ops.main.__call__)  # pyright  is quirky
+    type_test_dummy(ops.main.main)
+    type_test_negative(ops.main.__call__)  # type: ignore
+    type_test_negative(ops.main.main)  # type: ignore
+
+    ops.main(IdleCharm)
+
+    reset()
+    ops.main.main(IdleCharm)
+
+    with pytest.raises(TypeError):
+        ops.main()  # type: ignore
+
+    with pytest.raises(TypeError):
+        ops.main.main()  # type: ignore
+
+
+def test_submodule_import(reset: Reset):
+    import ops.main
+
+    type_test_dummy(ops.main.__call__)  # type: ignore FIXME
+    type_test_dummy(ops.main.main)
+    type_test_negative(ops.main.__call__)  # type: ignore
+    type_test_negative(ops.main.main)  # type: ignore
+
+    ops.main(IdleCharm)  # type: ignore FIXME
+
+    reset()
+    ops.main.main(IdleCharm)
+
+    with pytest.raises(TypeError):
+        ops.main()  # type: ignore
+
+    with pytest.raises(TypeError):
+        ops.main.main()  # type: ignore
+
+
+def test_import_from_top_level_module(reset: Reset):
+    from ops import main
+
+    type_test_dummy(main.__call__)
+    type_test_dummy(main.main)
+    type_test_negative(main.__call__)  # type: ignore
+    type_test_negative(main.main)  # type: ignore
+
+    main(IdleCharm)
+
+    reset()
```

Comment:
Instead of having a fixture that lets you do a reset like this, wouldn't it be simpler to just have two tests?

Reply:
To be honest the reason I grouped the test code like this was because I could not come up with a good way to name the individual tests. 🙈 

Would you have a good idea how to name these two distinctly? 🙏🏻 

```py
def test_top_level_import(reset: Reset):
    import ops

    ops.main(IdleCharm)


def test_top_level_import(reset: Reset):
    import ops

    ops.main.main(IdleCharm)
```

Reply:
done